snap/get_version

```diff
@@ -0,0 +1,116 @@
+set_version() {
+    version="${1}"
+
+    version=$(annotate_with_dirty "${version}")
+    craftctl set version="${version}"
+}
+
+# annotate_with_dirty may amend the version with a dirty markup if there are local changes.
+# $1: version: the version to annotate.
+annotate_with_dirty() {
+    version="${1}"
+
+    # check if current tree content is dirty.
+    is_dirty=$(git -C "${SNAPCRAFT_PART_SRC}" status --porcelain)
+    if [ -n "${is_dirty}" ]; then
+        version="${version}+dirty"
+    fi
+
+    echo "${version}"
+}
+
+# last_tag_on_branch returns the last tag on that branch that is not in another provided branch.
+# $1: origin_branch: branch we start listing from.
+# $2: other_branch: the branch where we don’t want to tags to be reachable from.
+# If we don’t have another branch to compare with, it will just return the most recent tag.
+last_tag_on_branch() {
+    origin_branch="${1}"
+    other_branch="${2}"
+
+    all_tags_on_origin_branch=$(git -C "${SNAPCRAFT_PART_SRC}" tag --sort=-committerdate --merged "${origin_branch}")
+    for tag in ${all_tags_on_origin_branch}; do
+        # Just return first commit if there is no branch to compare against the main branch.
+        if [ -z "${other_branch}" ]; then
+            tag=$(strip_branch_tag_prefix "${tag}" "${origin_branch}")
+            echo "${tag}"
+            return
+        fi
+
+        # Skip the tags that are reachable from the other branch.
+        commit_from_tag=$(git -C "${SNAPCRAFT_PART_SRC}" rev-list -n 1 "${tag}")
+        if git -C "${SNAPCRAFT_PART_SRC}" merge-base --is-ancestor "${commit_from_tag}" "${other_branch}"; then
+            continue
+        fi
+
+        tag=$(strip_branch_tag_prefix "${tag}" "${origin_branch}")
+        echo "${tag}"
+        return
+    done
+
+    echo "notag"
+}
+
+# strip_branch_tag_prefix will remove the branch name prefix from the tag name.
+# $1: tag: the tag name to strip the prefix from.
+# $2: current_branch: the branch name to strip from the tag.
+strip_branch_tag_prefix() {
+    tag="${1}"
+    current_branch="${2}"
+
+    echo "${tag#${current_branch}-}"
+}
+
+
+current_branch=$(git -C "${SNAPCRAFT_PART_SRC}" branch --show-current)
+
+# If current commit on that branch is tagged, this is the version we are using.
+current_commit_tag=$(git -C "${SNAPCRAFT_PART_SRC}" describe --exact-match --abbrev=0 --tags 2>/dev/null) || true
+if [ -n "${current_commit_tag}" ]; then
+    current_commit_tag=$(strip_branch_tag_prefix "${current_commit_tag}" "${current_branch}")
+    set_version "${current_commit_tag}"
+    exit 0
+fi
+
+# Current commit is not tagged, get commits sha.
+other_branch="main"
+if [ "${current_branch}" = "main" ]; then
+    other_branch=""
+fi
+
+# Get most recent tag on that branch not coming from the other branch.
+version=$(last_tag_on_branch "${current_branch}" "${other_branch}")
+
+# Append current commit sha.
+version="${version}+$(git -C "${SNAPCRAFT_PART_SRC}" describe --always)"
```

Comment:
This appends the tag if the current commit is tagged. I think we want to append `git rev-parse HEAD` instead.

Reply:
We don’t want to use `git rev-parse` which is giving a long sha and not a short one.

See my previous comment, if the current commit is tagged, we stopped. So the current commit is not tag and will thus append the short sha.

Reply:
>  if the current commit is tagged, we stopped. So the current commit is not tag

True.

We still don't want to append the output of `git describe` though, because even if the current commit is not a tag, it's not guaranteed to be the short commit SHA. If there is any tag on the branch (in a previous commit), the output is `${TAG}-${SHORT-COMMIT}`.

If we want the short commit SHA, we should use `git rev-parse --short HEAD` instead. I'm not sure we want that though, because the short SHA might not uniquely identify the commit.

Reply:
Unfortuntely, we have to use the short version. I’m happy to switch to this, even if I think this `${TAG}-${SHORT-COMMIT}` is only available if we use the `--all` or `--tags` options, and we don’t use them here.

Reply:
I can reproduce getting the `${TAG}-${SHORT-COMMIT}` output from `git describe --always`. So yeah, let's switch to `git rev-parse`. 

The length limit is unfortunate. I think we should specify the length of the SHA we want to include though, because else the length depends on how many commits the git repo has. I think we want the SHA to be 8 characters long, so that the whole version string doesn't exceed the 32 character limit. So we should use `git rev-parse --short=8 HEAD`.

Reply:
> I can reproduce getting the `${TAG}-${SHORT-COMMIT}` output from git describe --always. So yeah, let's switch to git rev-parse.

Waow, out of interest, can you give me reproducible steps for getting it? Learning everyday and we can learn something everyday :)

> The length limit is unfortunate. I think we should specify the length of the SHA we want to include though, because else the length depends on how many commits the git repo has. I think we want the SHA to be 8 characters long, so that the whole version string doesn't exceed the 32 character limit. So we should use git rev-parse --short=8 HEAD.

Right, but 8 characters still don’t cut it (see my last comment, with a "normal versionning" we would be at 31 characters due to dirty.

Or we can maybe change the "dirty" marker is it’s only for us and local builds anyway. Like just appending the `+` in the end (meaning it’s "this plus something"). What do you think?

Even if there is a clash, it means that then we can use our judgement as we have the tag we should looking from and even if possible, the chance of a clash for some informational content in such a short number of elements is negligeable and we could live with it IMHO. (But if we can go for 8 chars, let’s)

Reply:
> Waow, out of interest, can you give me reproducible steps for getting it?

Sure, for example on the main branch of my checkout of the adsys repo I get the following output:

```
$ git describe --always
0.8.4-1825-gda480813
```

> Right, but 8 characters still don’t cut it (see my last comment, with a "normal versionning" we would be at 31 characters due to dirty.

Why can't we go with 31 characters? The character limit for the version is 32, right?

If we need to use even fewer characters, we might want to consider dropping the commit from main. We can get it via `git merge-base main $COMMIT` where `$COMMIT` is the commit from the current branch. So it doesn't really give us any additional information but is "just" for developer convenience (which is a valid reason, I'm all for developer convenience :)).

Reply:
> Sure, for example on the main branch of my checkout of the adsys repo […]

Waow, indeed, there are definitively some black magic (well, it’s just code that we would need to have a look at :p) in term of I add the tag or not as it’s not in authd-oidc-brokers. Anyway, let’s use `rev-parse`.


> Why can't we go with 31 characters? The character limit for the version is 32, right?

You didn’t get me right, I meant, with the current situation in the example I gave with  7 characters in the [corresponding comment](https://github.com/ubuntu/authd-oidc-brokers/pull/161#discussion_r1766540899), this is 31 characters, so we can’t go with 8 characters for both current branch sha and main sha as this would be 33). Hence my proposal to simplify the `+dirty` suffix to just `+`. That way, we can even go up to 10 characters for each SHA.
We would then have: `24.09.3+d44131927.e6aac67a72+` which is 30 characters, marked as dirty and letting us having 2 digits for the patch version if the needs comes (which I hope will never happen). Does that sounds reasonable to you?

Reply:
> You didn’t get me right, I meant, with the current situation in the example I gave with 7 characters in the https://github.com/ubuntu/authd-oidc-brokers/pull/161#discussion_r1766540899, this is 31 characters

Ah ok, got it now.

> Hence my proposal to simplify the +dirty suffix to just +

Just appending `+` feels a bit too subtle to me. I don't want it to go unnoticed when there are changes in a snap version which are untracked. I would prefer going with 7 characters or dropping the commit from main.

Reply:
>> Hence my proposal to simplify the +dirty suffix to just +
> Just appending + feels a bit too subtle to me. I don't want it to go unnoticed when there are changes in a snap version which are untracked. I would prefer going with 7 characters or dropping the commit from main.

Ack, let’s forget about dropping dirty. I don’t really want to drop the commit from main either, if we have a general issue on brokers, this is a good way to ensure we know what level of code we have without digging through git history as we are mostly working from main.

So, if I get you right, you are ok to keep the current implementation with 7 characters for sha? (again, if we have collision, as we are between 2 tags, the chance of it should be really minor and it’s not a crucial info)

Reply:
This was what I meant. We are in agreement then!